### DISCUSSION TITLE
tagged-pointer-让指针包含更多信息

### DISCUSSION THREAD

**Discussion by caibingcheng** [caibingcheng/comment#305] [Announcements]
https://imcbc.cn/202211/tagged-pointer/

**Comment by penndu** [comment]
过来学习！

**Comment by caibingcheng** [comment]
> @penndu
> 过来学习！

凌晨一点半！(o_O) !

**Comment by lanyiyun666** [comment]
您好，博主
可以投稿服务器或者scdn文章内容吗？
这边可以赞助高防scdn【国内外均可】【一年起】
并且会给您一个心意红包
国内正规持证公司

联合计划介绍：https://www.tsyvps.com/bbs/?p=158

如果不方便发您的联系方式

以下是我的联系方式！

我的QQ是：2814841448

**Comment by panqiincs** [comment]
> @caibingcheng
> > @penndu
> > 过来学习！
> 
> 凌晨一点半！(o_O) !

是程序自动留言的么……

**Comment by caibingcheng** [comment]
> @panqiincs
> > @caibingcheng
> > > @penndu
> > > 过来学习！
> > 
> > 凌晨一点半！(o_O) !
> 
> 是程序自动留言的么……

因为真的没有睡

**Comment by penndu** [comment]
有段时间没更新了！

**Comment by caibingcheng** [comment]
> @penndu
> 有段时间没更新了！

(偷懒了...有些写一半先放下了)

**Comment by goodpaperman** [comment]
没明白，为何 aligned 到 4，地址低 2 位就留空了？

**Reply by goodpaperman** [reply]
而且一般指针已经 aligned 到 8 了吧

**Reply by caibingcheng** [reply]
指针多少字节对齐是不一定的呀，和平台、编译器、语言等都有关系。
比如按照8B对齐的平台，指定4B对齐也没问题吧，就是时间和空间的取舍。

---

align到4，地址低2bit为0。
比如一个地址0x0000，4B对齐的下一个偏移就是0x0004，对应的就是低2bit为0。
至于为什么不是0x0001到0x0005，我认为这种对齐方式应该也是可以的，因为字节对齐这件事是指按照内存起始位置算。
但是对于一般系统、平台，默认情况下都是4B、8B对齐，所以内存起始位置也是按照这些方式对齐，所以比如上文中制定的4B对齐，就会使得地址末两位为0。至于有没有3、5B对齐的平台，就超出我的知识范围了。
